### ISSUE TITLE
[stable/redis-ha] Use readinessProbe to control DNS routing of redis-ha-service

### ISSUE THREAD

**Issue by jackiechan1981** [helm/charts#16631]
Since all of our applications connect directly to the master node, we need redis-ha-service to always point to the master, whether or not the redis failover has occurred.
We try to use the readinessProbe, which is to use the readiness script to detect the role of the current redis node. If it is slave, which means it is not ready, redis-ha-service will not route to the slave node. Also redis-ha-service can be automatically routed to the new master node when redis failover has occurred.
I did some detailed tests and it works.
I wonder if this scheme is feasible?

The general configuration is as follows:
```yaml
readinessProbe:
  exec:
    command: ["sh", "/opt/appl/redis/probes/writeness.sh", "6379"]`
```

```shell
#!/bin/sh
set -eu
REDIS_HOST=localhost
REDIS_PORT=6379
REDIS_PASSWORD=pwd@redis

# check localhost redis instance`s role in pod, if is not master, then return 1, means not ready, so dns will not route to this instance
CHECK_SERVER="$(redis-cli -h "$REDIS_HOST" -p "$REDIS_PORT" -a "$REDIS_PASSWORD" role | grep master)"
if [ "$CHECK_SERVER" != "master" ]; then
  echo "Server readonly: $CHECK_SERVER"
  exit 1
fi
```

**Comment by DandyDeveloper** [comment]
@jackiechan1981 
This will likely be solved in my PR: https://github.com/helm/charts/pull/15305

This deploys a HAProxy that will effectively always point to master. 

That being said, the chart works as intended. Your sentinels are the utility you should be using for obtaining the address of the master, you should then use the library to connect to your Redis instance. 

Note, most libraries that support sentinel implicitly do this. 

Additional note, the current version of the chart assumes all communication is done from inside the k8 cluster.

**Comment by jackiechan1981** [comment]
Hi, @DandyDeveloper , thanks very much for your reply!

I just saw that this PR: #15305  has been approved and get merged into the master 4 hours ago, so it can be used directly!

I have been trying to solve this problem for a few days, until I found your great solution of HAProxy. Only because this PR has not been approved yet, so I haven't studied it in detail, but trying to find ohter solutions.

Thanks again for your reply and your great PR!
